src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultLSPDocumentManager.cs

```diff
@@ -0,0 +1,133 @@
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServerClient.Razor
+{
+    [Shared]
+    [Export(typeof(LSPDocumentManager))]
+    internal class DefaultLSPDocumentManager : LSPDocumentManagerBase
+    {
+        private readonly JoinableTaskContext _joinableTaskContext;
+        private readonly FileUriProvider _fileUriProvider;
+        private readonly LSPDocumentFactory _documentFactory;
+        private readonly Dictionary<Uri, DocumentTracker> _documents;
+
+        [ImportingConstructor]
+        public DefaultLSPDocumentManager(
+            JoinableTaskContext joinableTaskContext,
+            FileUriProvider fileUriProvider,
+            LSPDocumentFactory documentFactory)
+        {
+            if (joinableTaskContext is null)
+            {
+                throw new ArgumentNullException(nameof(joinableTaskContext));
+            }
+
+            if (fileUriProvider is null)
+            {
+                throw new ArgumentNullException(nameof(fileUriProvider));
+            }
+
+            if (documentFactory is null)
+            {
+                throw new ArgumentNullException(nameof(documentFactory));
+            }
+
+            _joinableTaskContext = joinableTaskContext;
+            _fileUriProvider = fileUriProvider;
+            _documentFactory = documentFactory;
+            _documents = new Dictionary<Uri, DocumentTracker>();
+        }
+
+        public override void TrackDocumentView(ITextBuffer buffer, ITextView textView)
+        {
+            if (buffer is null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (textView is null)
+            {
+                throw new ArgumentNullException(nameof(textView));
+            }
+
+            Debug.Assert(_joinableTaskContext.IsOnMainThread);
+
+            var uri = _fileUriProvider.GetOrCreate(buffer);
+            if (!_documents.TryGetValue(uri, out var documentTracker))
+            {
+                var lspDocument = _documentFactory.Create(buffer);
+                documentTracker = new DocumentTracker(lspDocument);
+                _documents[uri] = documentTracker;
+            }
+
+            documentTracker.TextViews.Add(textView);
+        }
+
+        public override void UntrackDocumentView(ITextBuffer buffer, ITextView textView)
+        {
+            if (buffer is null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (textView is null)
+            {
+                throw new ArgumentNullException(nameof(textView));
+            }
+
+            Debug.Assert(_joinableTaskContext.IsOnMainThread);
+
+            var uri = _fileUriProvider.GetOrCreate(buffer);
+            if (!_documents.TryGetValue(uri, out var documentTracker))
+            {
+                // We don't know about this document, noop.
+                return;
```

Comment:
>return [](start = 16, length = 6)

Nit: Would having an assert here make sense? Doesn't seem like this should be possible?

Reply:
OK, see your comment below about testing - maybe some sort of DEBUG-only Debug.Assert abstraction would make sense (that wouldn't actually Assert in unit tests, maybe log a message somewhere). Anyway, not a big deal. 

---
In reply to: [381416598](https://github.com/dotnet/aspnetcore-tooling/pull/1602#discussion_r381416598) [](ancestors = 381416598)

Reply:
> OK, see your comment below about testing - maybe some sort of DEBUG-only Debug.Assert abstraction would make sense (that wouldn't actually Assert in unit tests, maybe log a message somewhere). Anyway, not a big deal.

Ya I feel you there. I've built the code to work at runtime even in the unknown document case so it doesn't hurt to be extra protective. Going to leave as is in the effort of not investing too much time.